### PR TITLE
Respond with frontend config if there is no authenticated user call to /me

### PR DESCRIPTION
**Pull request by mayoral**
This PR implements #12790.

### CODE REVIEW

File: app/controllers/carto/api/users_controller.rb [hunk 1]
```diff
@@ -16,25 +16,35 @@ class UsersController < ::Api::ApplicationController
...
+        if current_viewer
+          carto_viewer = Carto::User.find(current_viewer.id)
+
+          response.merge!({
+            user_data: Carto::Api::UserPresenter.new(carto_viewer).data,
+            default_fallback_basemap: carto_viewer.default_basemap,
+            config: frontend_config_hash,
+            dashboard_notifications: carto_viewer.notifications_for_category(:dashboard),
+            is_just_logged_in: !!flash['logged'],
+            is_first_time_viewing_dashboard: !(carto_viewer.dashboard_viewed_at),
```
[the remark this answers is not in the document]
Comment (juanignaciosl) [reply app/controllers/carto/api/users_controller.rb]: ^ the message has room for improvement, but the parentheses can be safely removed.

File: app/controllers/carto/api/users_controller.rb [hunk 2]
```diff
@@ -16,25 +16,35 @@ class UsersController < ::Api::ApplicationController
...
-          is_first_time_viewing_dashboard: !(carto_viewer.dashboard_viewed_at),
-          user_frontend_version: carto_viewer.relevant_frontend_version,
-          asset_host: carto_viewer.asset_host
+          user_frontend_version: CartoDB::Application.frontend_version
         }
+
+        if current_viewer
+          carto_viewer = Carto::User.find(current_viewer.id)
+
+          response.merge!(
```
Comment (juanignaciosl) [inline app/controllers/carto/api/users_controller.rb]: Minor: the bang is not needed, as the result will be returned.

File: app/controllers/carto/api/users_controller.rb [hunk 3]
```diff
@@ -16,25 +16,35 @@ class UsersController < ::Api::ApplicationController
...
         render json: Carto::Api::UserPresenter.new(uri_user).data
       end
 
       def me
-        carto_viewer = Carto::User.find(current_viewer.id)
-
-        render json: {
-          user_data: Carto::Api::UserPresenter.new(carto_viewer).data,
-          default_fallback_basemap: carto_viewer.default_basemap,
+        response = {
```
Comment (juanignaciosl) [inline app/controllers/carto/api/users_controller.rb]: I think that this endpoint should always return the same structure, because of simplicity. If current viewer is not logged in most values would probably be `nil`, but the structure should be the same. What do you think?

Comment (mayoral) [reply app/controllers/carto/api/users_controller.rb]: Hmm... I see simpler in this way: If the user is authenticated, response gets enriched.

If we have to return the same format for both authenticated and unauthenticated, either we have to check field by field for user:

```
can_change_email: user.present? ? user.can_change_email? : nil
```

Or we have to write the hash two times, one with mostly nil fields, and the one with the authenticated fields.

File: app/controllers/carto/api/users_controller.rb (hunk 2 above)
Comment (mayoral) [reply app/controllers/carto/api/users_controller.rb]: 👍

File: app/controllers/carto/api/users_controller.rb (hunk 3 above)
Comment (juanignaciosl) [reply app/controllers/carto/api/users_controller.rb]: No, that's not what I meant. For the unauthenticated case we'd use a null object pattern, so the user would be an "anonymous" one. In your example, we'd get a `user` that has `can_change_email` set to `nil`.

I have mixed feelings. Summoning @javitonino as tie breaker.

Comment (javitonino) [reply app/controllers/carto/api/users_controller.rb]: I like @juanignaciosl approach, at least for this one. `user_data: nil` seems more explicit and easier to parse.

Also, it shouldn't be so hard, you can use `try`:
`user_frontend_version: carto_viewer.try(:relevant_frontend_version) || CartoDB::Application.frontend_version`
and similar. However, this is not as clear for attributes such as `is_first_time_viewing_dashboard`.

Anyway, I prefer the hash to be built in one place. I do not like the current approach of merging new data, overriding the previous values, since it forces you to look in two places to get the value that is being set. e.g: `user_frontend_version` is hard to follow.

Comment (mayoral) [reply app/controllers/carto/api/users_controller.rb]: Refactored to use try in https://github.com/CartoDB/cartodb/pull/12818/commits/489eb8c5f8cb3eba55c3c3fbbbae1492790a362c

File: app/controllers/carto/api/users_controller.rb [hunk 4]
```diff
@@ -16,24 +16,25 @@ class UsersController < ::Api::ApplicationController
...
+      before_filter :optional_api_authorization, only: [:me]
+      skip_before_filter :api_authorization_required, only: [:me, :get_authenticated_users]
 
       def show
         render json: Carto::Api::UserPresenter.new(uri_user).data
       end
 
       def me
-        carto_viewer = Carto::User.find(current_viewer.id)
+        carto_viewer = current_viewer.present? ? Carto::User.find(current_viewer.id) : nil
```
Comment (juanignaciosl) [inline app/controllers/carto/api/users_controller.rb]: Post-CR comment: next time you have to edit this, you'd probably want to return a "null user" here, a constant instance of `Carto::User` that contains the default values for the anonymous user. That's a clean way to specify defaults, instead of using ternaries everywhere.
